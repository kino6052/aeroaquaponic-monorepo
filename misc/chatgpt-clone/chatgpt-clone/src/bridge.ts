import { Decoupler } from "../../../../utils/Decoupler";

export const decoupler = new Decoupler({}, (state: {}) => ({}));
