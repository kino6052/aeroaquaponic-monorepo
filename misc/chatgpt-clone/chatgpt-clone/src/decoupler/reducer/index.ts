import { EModel, EUser } from "../enums";
import {
  EActionType,
  EControlId,
  IState,
  TAction,
  TConversation,
  TConversationCategory,
  TMainProps,
  TMessage,
} from "../types";

import { uuid } from "uuidv4";

export const selectMainProps = (state: IState): TMainProps => {
  return {
    input: state.input,
    messages: state.messages,
    activeMessage: state.activeMessage,
    conversations: state.conversations,
    isOpen: state.isOpen,
    selectedModel: state.selectedModel,
  };
};

export function reducer<T>(state: IState, action: TAction<T>): IState {
  switch (action.id.id) {
    case EControlId.ExpandButton:
      return {
        ...state,
        isOpen: !state.isOpen,
      };

    case EControlId.QueryInput:
      if (action.type !== EActionType.Change) return state;
      return {
        ...state,
        input: action.payload as string,
      };

    case EControlId.Toggle:
      return {
        ...state,
        selectedModel:
          state.selectedModel === EModel.GPT3 ? EModel.GPT4 : EModel.GPT3,
      };

    case EControlId.Submit:
      const newConversation: TConversation = {
        id: uuid(),
        name: "Conversation " + (state.conversations.length + 1),
        isActive: true,
      };

      const newCategory: TConversationCategory = {
        id: uuid(),
        category: "Today",
        conversations: [newConversation],
      };

      const newMessage: TMessage = {
        user: EUser.User,
        text: state.input,
      };

      return {
        ...state,
        input: "",
        activeMessage: "",
        messages: [...state.messages, newMessage],
        conversations: [...state.conversations, newCategory],
      };

    default:
      return state;
  }
}
