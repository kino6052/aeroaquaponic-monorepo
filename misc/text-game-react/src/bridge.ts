import { BehaviorSubject, Subject } from "rxjs";
import { EntityMap } from "./reduce/model/entities";
import * as outputs from "./reduce/outputs/outputs";

export enum Id {
  Input = "Input",
}

export type TAvailableCommand = {
  name: string;
  description: string;
  args: TAvailableCommand[];
};

export interface IBrowser {
  current?: string;
  visited: string[];
  resources: {
    [key: string]: {
      url: string;
      description: string;
    };
  };
}

export type TEntityType = "world" | "quest" | "objective" | "cli" | "misc";

export interface SerializedEntity {
  id: string;
  entities: string[];
  type: TEntityType;
  name: string;
  description: string;
}

export interface IState {
  input: string;
  output: string;
  history: string[];
  entities: { [key: string]: SerializedEntity };
}

export type TCommand = "change" | "enter" | "suggest" | "load";

export type TEvent = [TCommand, string];

export const AppEventSubject = new Subject<TEvent>();

export const initialState: IState = {
  input: "",
  history: [],
  output: outputs.initialOutput,
  entities: EntityMap,
};

export const StateSubject = new BehaviorSubject<IState>(initialState);
