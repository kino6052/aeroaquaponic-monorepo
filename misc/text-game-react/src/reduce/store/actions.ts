import { TEvent } from "../../bridge";

export const getChangeAction = (input: string): TEvent => ["change", input];
export const getEnterAction = (): TEvent => ["enter", ""];
export const getSuggestAction = (): TEvent => ["suggest", ""];
