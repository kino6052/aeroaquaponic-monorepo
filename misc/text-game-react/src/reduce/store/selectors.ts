import { Id, IState } from "../../bridge";

export const selectCurrent = (state: IState) => state.currentId;
export const selectHistory = (state: IState) => state.history;
export const selectNext = (state: IState) => state.next;
export const selectChildren = (state: IState, id: Id) =>
  state.data[id]?.children || [];
