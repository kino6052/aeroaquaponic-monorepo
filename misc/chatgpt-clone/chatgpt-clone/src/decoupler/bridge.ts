import { Decoupler } from "../../utils/Decoupler";
import { EModel } from "./enums";
import { reducer } from "./reducer";
import { EActionType, EControlId, IState } from "./types";

export const initialState: IState = {
  input: "",
  messages: [],
  activeMessage: "",
  conversations: [],
  isOpen: true,
  selectedModel: EModel.GPT3,
};

export const decoupler = new Decoupler<IState, EActionType, EControlId, string>(
  initialState,
  reducer
);
