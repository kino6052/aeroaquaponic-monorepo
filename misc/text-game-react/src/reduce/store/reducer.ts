import produce from "immer";
import { Id, IState, TEvent } from "../../bridge";
import { selectChildren } from "./selectors";

export const reduce = (event: TEvent, state: IState): IState => {
  return produce(state, (draft) => {
    if (event[0] === "restore") {
      const query = event[1] || "";
      const result = query.replace("?", "").split("&");
      draft.history = result.slice(0, -1) as Id[];
      const currentId = result.slice(-1)[0] as Id | undefined;
      if (currentId) {
        draft.currentId = currentId;
        draft.next = selectChildren(draft, currentId);
      }
      return;
    }
    if (event[0] === "select") {
      if (draft.next.includes(event[1] as Id)) {
        const id = event[1] as Id;
        draft.currentId && draft.history.push(draft.currentId);
        draft.currentId = id;
        draft.next = draft.data[id].children;
        return;
      }
      if (draft.history.includes(event[1] as Id)) {
        const id = event[1] as Id;
        draft.next = draft.data[id].children;
        const index = draft.history.findIndex((_id) => _id === id);
        draft.history = draft.history.slice(0, index);
        draft.currentId = id;
        return;
      }
      return;
    }
  });
};

export const compose = (state: IState) => (events: TEvent[]) =>
  events.reduce((_state, event) => reduce(event, _state), state);
