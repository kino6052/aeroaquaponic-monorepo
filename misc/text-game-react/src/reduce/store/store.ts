import { IState, TAvailableCommand } from "../../bridge";
import { makeHeader, makeList, makeParagraph } from "../utils";
import { selectCommands } from "./selectors";

export const commands: Record<string, TAvailableCommand> = {
  status: {
    name: "status",
    description: "Find out what's going on at the moment",
    args: [],
  },
  help: {
    name: "help",
    description: "Find out what available commands are there",
    args: [],
  },
  google: {
    name: "google",
    description: "Google search",
    args: [
      {
        name: "Self-sufficiency",
        description: "Learn about self-sufficiency",
        args: [],
      },
    ],
  },
};

export const getCommandData = (
  state: IState,
  commandName?: typeof commands[number]["name"]
): { args: string[]; name?: string; description?: string } => {
  const commands = selectCommands(state);
  const command = commands[commandName || ""];
  if (!command)
    return {
      args: Object.keys(commands),
      name: "Available commands",
      description: `
Note: You can autocomplete queries by hitting Tab. For example, enter "goo" and hit Tab key, you will get "google"`,
    };
  const { args, description, name } = command;
  return {
    args: args.map(({ name }) => name),
    description: description ?? "",
    name: name ?? "",
  };
};

const _generateCommandOutput = ({
  args,
  description,
  name,
}: ReturnType<typeof getCommandData>) => `
${makeHeader(name ?? "")}
${makeParagraph(description ?? "")}
${makeList("", args)}
`;

export const generateCommandOutput = (
  state: IState,
  commandName?: typeof commands[number]["name"]
) => _generateCommandOutput(getCommandData(state, commandName));
