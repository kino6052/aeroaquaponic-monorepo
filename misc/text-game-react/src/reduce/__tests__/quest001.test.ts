import { initialState } from "../../bridge";
import {
  getChangeAction,
  getEnterAction,
  getSuggestAction,
} from "../store/actions";
import { compose } from "../store/reducer";
import { selectHistory, selectInput, selectOutput } from "../store/selectors";

describe("Quest 001", () => {
  it("should have a todo item", () => {
    const resultingState = compose(initialState)([
      getChangeAction("todo"),
      getEnterAction(),
    ]);
    expect(selectInput(resultingState)).toEqual("");
    expect(selectOutput(resultingState)).toMatchInlineSnapshot(
      `"<h3>Todo</h3><p>I had a look at my todo and here were the items:</p><p>Here is what's left: </p><ul><li>Go to the internet to visit the site and learn about self-sufficiency</li></ul>"`
    );
  });

  it("should have a correct autocomplete sequence", () => {
    const resultingState = compose(initialState)([
      getChangeAction("internet self"),
      getSuggestAction(),
      getEnterAction(),
    ]);
    expect(selectInput(resultingState)).toEqual("");
    expect(selectHistory(resultingState).length).toMatchInlineSnapshot(`1`);
    expect(selectOutput(resultingState)).toMatchInlineSnapshot(
      `"<h3>self-sufficiency.com</h3><p>I visited the website about self-sufficiency that I had come across a few days ago. It was an eye-opening experience, and I learned a lot about what it takes to truly become self-sufficient. The website explained that in order to start the journey, I need to form a unit of self-sufficiency, which comprises of three main components: property, technology, and skills.</p><p>The first component, property, is crucial. I need to find a piece of land where I can live without paying rent and where I can grow my food. This will form the foundation of my self-sufficiency journey, and it's something I'll need to start researching and looking into.</p><p>The second component, technology, is also crucial. Over time, I'll need to improve my property and make sure that I don't spend too much time doing unnecessary repetitive labor. This means investing in tools and equipment that will make my life easier and more efficient. It's important to find the balance between working hard and living a well-balanced life.</p><p>The third component, skills, is also key. I need to start developing skills that will allow me to adapt quickly to any change in the world, and not let me lose the self-sufficient status. This means learning new things, experimenting with different techniques, and being open to new ideas.</p><p>Overall, it was an enlightening experience, and it gave me a lot to think about. I now have a better understanding of what it takes to become truly self-sufficient, and I'm excited to start researching and taking action on the different components. I'm looking forward to updating my diary with my progress and thoughts as I learn more about self-sufficiency.</p>"`
    );
  });

  it("should have a todo item", () => {
    const resultingState = compose(initialState)([
      getChangeAction("internet self-sufficiency"),
      getEnterAction(),
      getChangeAction("todo"),
      getEnterAction(),
    ]);
    expect(selectInput(resultingState)).toEqual("");
    expect(selectOutput(resultingState)).toMatchInlineSnapshot(
      `"<h3>Todo</h3><p>I had a look at my todo and here were the items:</p><p>Here is what's left: </p><ul><li>go to the internet and find out about land</li></ul>"`
    );
  });

  it("should only display message once", () => {
    const resultingState = compose(initialState)([
      getChangeAction("internet self-sufficiency"),
      getEnterAction(),
      getChangeAction("internet self-sufficiency"),
      getEnterAction(),
    ]);
    expect(selectInput(resultingState)).toEqual("");
    expect(selectOutput(resultingState)).toMatchInlineSnapshot(
      `"<h3>self-sufficiency.com</h3><p>I really dug the ideas outlined in the website about self-sufficiency.</p><p>The concept of forming a unit of self-sufficiency really resonated with me.</p><p>Now I just need to follow my todo list</p>"`
    );
  });
});
