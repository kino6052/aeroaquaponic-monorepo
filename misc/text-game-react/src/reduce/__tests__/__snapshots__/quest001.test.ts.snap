// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quest 001 should have phone numbers after visiting the website 1`] = `"phone"`;
